Creating a working Window XP binary

**OS:** Windows XP SP3 Pro 32bit  
**Version:** 0.2.4

I've read that you don't support WinXP anymore, but for a friend I've tried to cross-compile v0.2.4. nonetheless. I'm using MingGW-w64 6.0.0 and GCC 9.2.0.  

Building all the prerequisites went just fine.  
First time compiling OpenRCT2 I got:
```
[  0%] Building CXX object CMakeFiles/libopenrct2.dir/src/openrct2/Cheats.cpp.obj
[  0%] Building CXX object CMakeFiles/libopenrct2.dir/src/openrct2/CmdlineSprite.cpp.obj
[  0%] Building CXX object CMakeFiles/libopenrct2.dir/src/openrct2/Context.cpp.obj
[...]
/cygdrive/m/[...]/mingw-w64-i686/i686-w64-mingw32/include/c++/9.2.0/tuple:185:12: error: invalid use of incomplete type 'class std::promise<void>'
  185 |     struct _Tuple_impl<_Idx, _Head, _Tail...>
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

By adding `#include "mingw.future.h"` (from [mingw-std-threads](https://github.com/meganz/mingw-std-threads)) to `src/openrct2/Context.h` this was fixed.  
Now I'm getting the following error:
```
[...]
[  8%] Building CXX object CMakeFiles/libopenrct2.dir/src/openrct2/config/Config.cpp.obj
/cygdrive/m/[...]/OpenRCT2-0.2.4/src/openrct2/config/Config.cpp: In function 'bool Config::WriteFile(const string&)':
/cygdrive/m/[...]/OpenRCT2-0.2.4/src/openrct2/config/Config.cpp:588:19: error: 'CreateDirectory' is not a member of 'Path'; did you mean 'CreateDirectoryA'?
  588 |             Path::CreateDirectory(directory);
      |                   ^~~~~~~~~~~~~~~
      |                   CreateDirectoryA
make[2]: *** [CMakeFiles/libopenrct2.dir/build.make:554: CMakeFiles/libopenrct2.dir/src/openrct2/config/Config.cpp.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:163: CMakeFiles/libopenrct2.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

If this is a WinXP specific error, then this issue can of course be closed. If this is a project specific error on the other hand, can it be fixed?

Thank you.

> Try placing `#undef CreateDirectory` at the top of the path.hpp or config.cpp after windows headers.

Thanks. That did the trick.

Next I replaced every...
```
#include <thread>
#include <mutex>
#include <future>
```

...in every file in the repo with...
```
#include "mingw.thread.h"
#include "mingw.mutex.h"
#include "mingw.future.h"
```
...to prevent more errors like...

```
error: 'thread' is not a member of 'std'
error: 'mutex' in namespace 'std' does not name a type
error: 'mutex' is not a member of 'std'
error: invalid use of incomplete type 'class std::promise<void>'
```

With that settled I'm now getting the following error:
```
[ 38%] Building CXX object CMakeFiles/libopenrct2.dir/src/openrct2/platform/Platform.Win32.cpp.obj
/cygdrive/m/[...]/OpenRCT2-0.2.4/src/openrct2/platform/Platform.Win32.cpp: In function 'bool Platform::IsOSVersionAtLeast(uint32_t, uint32_t, uint32_t)':
/cygdrive/m/[...]/OpenRCT2-0.2.4/src/openrct2/platform/Platform.Win32.cpp:217:80: error: cast between incompatible function types from 'FARPROC' {aka 'int (__attribute__((stdcall)) *)()'} to 'RtlGetVersionPtr' {aka 'long int (__attribute__((stdcall)) *)(_OSVERSIONINFOW*)'} [-Werror=cast-function-type]
  217 |             auto fn = (RtlGetVersionPtr)GetProcAddress(hModule, "RtlGetVersion");
      |                                                                                ^
```

Would this be a WinXP specific error?

> OpenRCT2 builds just fine for winnt 5.1 with no changes to source necessary: https://github.com/janisozaur/openrct2-mingw-winnt5.1

Your repo doesn't look at all like this repo. Where is the `src` dir for instance?

> You could create a simple library with those two symbols and use `LD_PRELOAD`-type mechanism to ensure OpenRCT2 can use them.

Unfortunately I'm no C/C++ coder. I don't know how.

Alright, I had another go at this, but there's first of all something I forgot to mention last time.

---

There's something going wrong with SDL2's dependencies. Despite correctly being detected by `cmake`,...

`build/CMakeCache.txt`:
```
[...]
SDL2_STATIC_LDFLAGS:INTERNAL=-L/cygdrive/[...]/mingw-w64-i686/i686-w64-mingw32/lib;-lmingw32;-lSDL2main;-lSDL2;-lmingw32;-lSDL2main;-lSDL2;-Wl,--no-undefined;-Wl,--dynamicbase;-Wl,--nxcompat;-lm;-ldinput8;-ldxguid;-ldxerr8;-luser32;-lgdi32;-lwinmm;-limm32;-lole32;-loleaut32;-lshell32;-lsetupapi;-lversion;/lib/w32api/libuuid.a;-static-libgcc
[...]
```

...the moment `openrct2.exe` is being created I get errors like:
```
undefined reference to `_imp__SetupDiGetClassDevsA@16'
undefined reference to `_imp__SetupDiEnumDeviceInterfaces@20'
undefined reference to `_imp__SetupDiGetDeviceInterfaceDetailA@24'
```

`build/CMakeFiles/openrct2.dir/linklibs.rsp`, which I believe is a list of all the libraries that need to be linked to `libopenrct2.a`, somehow didn't have the crucial `-lsetupapi`!

```bash
sed -i "s/-lSDL2 /&-lsetupapi /" build/CMakeFiles/openrct2.dir/linklibs.rsp
```
After `cmake` this `sed`-line fixed that for me.

---

So far I had to apply the following patches to fix quite some errors:

`openrct2_mingw-std-threads.patch` (quote from https://github.com/meganz/mingw-std-threads: _"Implementation of standard C++11 threading classes, which are currently still missing on MinGW GCC."_):
```diff
diff --git a/src/openrct2-ui/drawing/engines/opengl/TextureCache.h b/src/openrct2-ui/drawing/engines/opengl/TextureCache.h
index 5ee5a74..9f82811 100644
--- a/src/openrct2-ui/drawing/engines/opengl/TextureCache.h
+++ b/src/openrct2-ui/drawing/engines/opengl/TextureCache.h
@@ -15,10 +15,10 @@
 #include <SDL_pixels.h>
 #include <algorithm>
 #include <array>
-#include <mutex>
+#include "mingw.mutex.h"
 #include <openrct2/common.h>
 #ifndef __MACOSX__
-#    include <shared_mutex>
+#    include "mingw.shared_mutex.h"
 #endif
 #include <unordered_map>
 #include <vector>
diff --git a/src/openrct2-ui/windows/ObjectLoadError.cpp b/src/openrct2-ui/windows/ObjectLoadError.cpp
index acf6163..54a15fe 100644
--- a/src/openrct2-ui/windows/ObjectLoadError.cpp
+++ b/src/openrct2-ui/windows/ObjectLoadError.cpp
@@ -7,7 +7,7 @@
  * OpenRCT2 is licensed under the GNU General Public License version 3.
  *****************************************************************************/
 
-#include <mutex>
+#include "mingw.mutex.h"
 #include <openrct2-ui/interface/Widget.h>
 #include <openrct2-ui/windows/Window.h>
 #include <openrct2/Context.h>
diff --git a/src/openrct2/core/Http.cURL.cpp b/src/openrct2/core/Http.cURL.cpp
index 3a1f690..78c1775 100644
--- a/src/openrct2/core/Http.cURL.cpp
+++ b/src/openrct2/core/Http.cURL.cpp
@@ -16,7 +16,7 @@
 #    include <cstring>
 #    include <memory>
 #    include <stdexcept>
-#    include <thread>
+#    include "mingw.thread.h"
 
 #    ifdef _WIN32
 // cURL includes windows.h, but we don't need all of it.
diff --git a/src/openrct2/core/Http.h b/src/openrct2/core/Http.h
index 969446a..8205709 100644
--- a/src/openrct2/core/Http.h
+++ b/src/openrct2/core/Http.h
@@ -16,7 +16,7 @@
 #    include <functional>
 #    include <map>
 #    include <string>
-#    include <thread>
+#    include "mingw.thread.h"
 
 namespace Http
 {
diff --git a/src/openrct2/core/JobPool.hpp b/src/openrct2/core/JobPool.hpp
index 4550f5f..f199f53 100644
--- a/src/openrct2/core/JobPool.hpp
+++ b/src/openrct2/core/JobPool.hpp
@@ -12,11 +12,11 @@
 #include <algorithm>
 #include <atomic>
 #include <cassert>
-#include <condition_variable>
+#include "mingw.condition_variable.h"
 #include <deque>
 #include <functional>
-#include <mutex>
-#include <thread>
+#include "mingw.mutex.h"
+#include "mingw.thread.h"
 #include <vector>
 
 class JobPool
diff --git a/src/openrct2/drawing/ScrollingText.cpp b/src/openrct2/drawing/ScrollingText.cpp
index d06dd64..0afd1dc 100644
--- a/src/openrct2/drawing/ScrollingText.cpp
+++ b/src/openrct2/drawing/ScrollingText.cpp
@@ -17,7 +17,7 @@
 #include "TTF.h"
 
 #include <algorithm>
-#include <mutex>
+#include "mingw.mutex.h"
 
 struct rct_draw_scroll_text
 {
diff --git a/src/openrct2/drawing/TTF.cpp b/src/openrct2/drawing/TTF.cpp
index b7d2a07..2d92df4 100644
--- a/src/openrct2/drawing/TTF.cpp
+++ b/src/openrct2/drawing/TTF.cpp
@@ -10,7 +10,7 @@
 #ifndef NO_TTF
 
 #    include <atomic>
-#    include <mutex>
+#    include "mingw.mutex.h"
 #    pragma clang diagnostic push
 #    pragma clang diagnostic ignored "-Wdocumentation"
 #    include <ft2build.h>
diff --git a/src/openrct2/interface/InteractiveConsole.cpp b/src/openrct2/interface/InteractiveConsole.cpp
index 85eba8f..d373c4c 100644
--- a/src/openrct2/interface/InteractiveConsole.cpp
+++ b/src/openrct2/interface/InteractiveConsole.cpp
@@ -54,7 +54,7 @@
 #include <deque>
 #include <exception>
 #include <string>
-#include <thread>
+#include "mingw.thread.h"
 #include <vector>
 
 #ifndef NO_TTF
diff --git a/src/openrct2/interface/InteractiveConsole.h b/src/openrct2/interface/InteractiveConsole.h
index ec1a318..ca43322 100644
--- a/src/openrct2/interface/InteractiveConsole.h
+++ b/src/openrct2/interface/InteractiveConsole.h
@@ -13,7 +13,7 @@
 #include "../localisation/FormatCodes.h"
 
 #include <deque>
-#include <future>
+#include "mingw.future.h"
 #include <queue>
 #include <string>
 #include <tuple>
diff --git a/src/openrct2/network/ServerList.h b/src/openrct2/network/ServerList.h
index 75bff4e..b6882ab 100644
--- a/src/openrct2/network/ServerList.h
+++ b/src/openrct2/network/ServerList.h
@@ -11,7 +11,7 @@
 
 #include "../common.h"
 
-#include <future>
+#include "mingw.future.h"
 #include <optional>
 #include <stdexcept>
 #include <string>
diff --git a/src/openrct2/network/Socket.cpp b/src/openrct2/network/Socket.cpp
index 6ecda08..5f1802c 100644
--- a/src/openrct2/network/Socket.cpp
+++ b/src/openrct2/network/Socket.cpp
@@ -13,9 +13,9 @@
 #    include <chrono>
 #    include <cmath>
 #    include <cstring>
-#    include <future>
+#    include "mingw.future.h"
 #    include <string>
-#    include <thread>
+#    include "mingw.thread.h"
 
 // clang-format off
 // MSVC: include <math.h> here otherwise PI gets defined twice
diff --git a/src/openrct2/object/ObjectManager.cpp b/src/openrct2/object/ObjectManager.cpp
index 0cbe3ca..5165a2e 100644
--- a/src/openrct2/object/ObjectManager.cpp
+++ b/src/openrct2/object/ObjectManager.cpp
@@ -26,8 +26,8 @@
 #include <algorithm>
 #include <array>
 #include <memory>
-#include <mutex>
-#include <thread>
+#include "mingw.mutex.h"
+#include "mingw.thread.h"
 #include <unordered_set>
 
 class ObjectManager final : public IObjectManager
```

`openrct2_undef-CreateDirectory.patch` (see [issuecomment-592982457](https://github.com/OpenRCT2/OpenRCT2/issues/10850#issuecomment-592982457)):
```diff
diff --git a/src/openrct2/core/Path.hpp b/src/openrct2/core/Path.hpp
index 924b269..2529156 100644
--- a/src/openrct2/core/Path.hpp
+++ b/src/openrct2/core/Path.hpp
@@ -12,6 +12,7 @@
 #include "../common.h"
 
 #include <string>
+#undef CreateDirectory
 
 namespace Path
 {
```

`openrct2_suppres-FARPROC-warning.patch`:
```diff
diff --git a/src/openrct2/platform/Platform.Win32.cpp b/src/openrct2/platform/Platform.Win32.cpp
index 5c095f0..1d04e32 100644
--- a/src/openrct2/platform/Platform.Win32.cpp
+++ b/src/openrct2/platform/Platform.Win32.cpp
@@ -214,7 +214,14 @@ namespace Platform
         if (hModule != nullptr)
         {
             using RtlGetVersionPtr = NTSTATUS(WINAPI*)(PRTL_OSVERSIONINFOW);
-            auto fn = (RtlGetVersionPtr)GetProcAddress(hModule, "RtlGetVersion");
+#           if defined(__GNUC__)
+#           pragma GCC diagnostic push
+#           pragma GCC diagnostic ignored "-Wcast-function-type"
+#           endif
+                auto fn = (RtlGetVersionPtr)GetProcAddress(hModule, "RtlGetVersion");
+#           if defined(__GNUC__)
+#           pragma GCC diagnostic pop
+#           endif
             if (fn != nullptr)
             {
                 RTL_OSVERSIONINFOW rovi{};

```

`openrct2_winxp-friendly-functions.patch`:
```diff
diff --git a/src/openrct2/platform/Crash.cpp b/src/openrct2/platform/Crash.cpp
index fef8f25..cd7afaf 100644
--- a/src/openrct2/platform/Crash.cpp
+++ b/src/openrct2/platform/Crash.cpp
@@ -108,18 +108,18 @@ static bool OnCrash(
     wchar_t saveFilePath[MAX_PATH];
     wchar_t configFilePath[MAX_PATH];
     wchar_t saveFilePathGZIP[MAX_PATH];
-    swprintf_s(dumpFilePath, std::size(dumpFilePath), L"%s\\%s.dmp", dumpPath, miniDumpId);
-    swprintf_s(saveFilePath, std::size(saveFilePath), L"%s\\%s.sv6", dumpPath, miniDumpId);
-    swprintf_s(configFilePath, std::size(configFilePath), L"%s\\%s.ini", dumpPath, miniDumpId);
-    swprintf_s(saveFilePathGZIP, std::size(saveFilePathGZIP), L"%s\\%s.sv6.gz", dumpPath, miniDumpId);
+    swprintf(dumpFilePath, std::size(dumpFilePath), L"%s\\%s.dmp", dumpPath, miniDumpId);
+    swprintf(saveFilePath, std::size(saveFilePath), L"%s\\%s.sv6", dumpPath, miniDumpId);
+    swprintf(configFilePath, std::size(configFilePath), L"%s\\%s.ini", dumpPath, miniDumpId);
+    swprintf(saveFilePathGZIP, std::size(saveFilePathGZIP), L"%s\\%s.sv6.gz", dumpPath, miniDumpId);
 
     wchar_t dumpFilePathNew[MAX_PATH];
-    swprintf_s(
+    swprintf(
         dumpFilePathNew, std::size(dumpFilePathNew), L"%s\\%s(%s_%s).dmp", dumpPath, miniDumpId, _wszCommitSha1Short,
         _wszArchitecture);
 
     wchar_t dumpFilePathGZIP[MAX_PATH];
-    swprintf_s(dumpFilePathGZIP, std::size(dumpFilePathGZIP), L"%s.gz", dumpFilePathNew);
+    swprintf(dumpFilePathGZIP, std::size(dumpFilePathGZIP), L"%s.gz", dumpFilePathNew);
 
     // Compress the dump
     {
@@ -215,7 +215,7 @@ static bool OnCrash(
                                              L"We would like to upload the crash dump for automated analysis, do you agree?\n"
                                              L"The automated analysis is done by courtesy of https://backtrace.io/";
     wchar_t message[MAX_PATH * 2];
-    swprintf_s(message, MessageFormat, dumpFilePath, WSZ(OPENRCT2_VERSION), _wszCommitSha1Short);
+    swprintf(message, MessageFormat, dumpFilePath, WSZ(OPENRCT2_VERSION), _wszCommitSha1Short);
 
     // Cannot use platform_show_messagebox here, it tries to set parent window already dead.
     int answer = MessageBoxW(nullptr, message, WSZ(OPENRCT2_NAME), MB_YESNO | MB_ICONERROR);
@@ -231,7 +231,7 @@ static bool OnCrash(
                                             L"Please provide following information as well:\n"
                                             L"Error code = %d\n"
                                             L"Response = %s";
-            swprintf_s(message, MessageFormat2, error, response.c_str());
+            swprintf(message, MessageFormat2, error, response.c_str());
             MessageBoxW(nullptr, message, WSZ(OPENRCT2_NAME), MB_OK | MB_ICONERROR);
         }
         else
diff --git a/src/openrct2/platform/Windows.cpp b/src/openrct2/platform/Windows.cpp
index 5117767..57639f9 100644
--- a/src/openrct2/platform/Windows.cpp
+++ b/src/openrct2/platform/Windows.cpp
@@ -410,7 +410,7 @@ uint8_t platform_get_locale_date_format()
     //
     wchar_t first[sizeof(dateFormat)];
     wchar_t second[sizeof(dateFormat)];
-    if (swscanf_s(
+    if (swscanf(
             dateFormat, L"%l[dyM]%*l[^dyM]%l[dyM]%*l[^dyM]%*l[dyM]", first, (uint32_t)std::size(first), second,
             (uint32_t)std::size(second))
         != 2)
@@ -547,203 +547,4 @@ bool platform_process_is_elevated()
 }
 
 ///////////////////////////////////////////////////////////////////////////////
-// File association setup
-///////////////////////////////////////////////////////////////////////////////
-
-#    define SOFTWARE_CLASSES L"Software\\Classes"
-#    define MUI_CACHE L"Local Settings\\Software\\Microsoft\\Windows\\Shell\\MuiCache"
-
-static std::wstring get_progIdName(const std::string_view& extension)
-{
-    auto progIdName = std::string(OPENRCT2_NAME) + std::string(extension);
-    auto progIdNameW = String::ToWideChar(progIdName);
-    return progIdNameW;
-}
-
-static bool windows_setup_file_association(
-    const utf8* extension, const utf8* fileTypeText, const utf8* commandText, const utf8* commandArgs, const uint32_t iconIndex)
-{
-    wchar_t exePathW[MAX_PATH];
-    wchar_t dllPathW[MAX_PATH];
-
-    [[maybe_unused]] int32_t printResult;
-
-    GetModuleFileNameW(nullptr, exePathW, (DWORD)std::size(exePathW));
-    GetModuleFileNameW(plaform_get_dll_module(), dllPathW, (DWORD)std::size(dllPathW));
-
-    auto extensionW = String::ToWideChar(extension);
-    auto fileTypeTextW = String::ToWideChar(fileTypeText);
-    auto commandTextW = String::ToWideChar(commandText);
-    auto commandArgsW = String::ToWideChar(commandArgs);
-    auto progIdNameW = get_progIdName(extension);
-
-    bool result = false;
-    HKEY hKey = nullptr;
-    HKEY hRootKey = nullptr;
-
-    // [HKEY_CURRENT_USER\Software\Classes]
-    if (RegOpenKeyW(HKEY_CURRENT_USER, SOFTWARE_CLASSES, &hRootKey) != ERROR_SUCCESS)
-    {
-        goto fail;
-    }
-
-    // [hRootKey\.ext]
-    if (RegSetValueW(hRootKey, extensionW.c_str(), REG_SZ, progIdNameW.c_str(), 0) != ERROR_SUCCESS)
-    {
-        goto fail;
-    }
-
-    if (RegCreateKeyW(hRootKey, progIdNameW.c_str(), &hKey) != ERROR_SUCCESS)
-    {
-        goto fail;
-    }
-
-    // [hRootKey\OpenRCT2.ext]
-    if (RegSetValueW(hKey, nullptr, REG_SZ, fileTypeTextW.c_str(), 0) != ERROR_SUCCESS)
-    {
-        goto fail;
-    }
-    // [hRootKey\OpenRCT2.ext\DefaultIcon]
-    wchar_t szIconW[MAX_PATH];
-    printResult = swprintf_s(szIconW, MAX_PATH, L"\"%s\",%d", dllPathW, iconIndex);
-    assert(printResult >= 0);
-    if (RegSetValueW(hKey, L"DefaultIcon", REG_SZ, szIconW, 0) != ERROR_SUCCESS)
-    {
-        goto fail;
-    }
-
-    // [hRootKey\OpenRCT2.sv6\shell]
-    if (RegSetValueW(hKey, L"shell", REG_SZ, L"open", 0) != ERROR_SUCCESS)
-    {
-        goto fail;
-    }
-
-    // [hRootKey\OpenRCT2.sv6\shell\open]
-    if (RegSetValueW(hKey, L"shell\\open", REG_SZ, commandTextW.c_str(), 0) != ERROR_SUCCESS)
-    {
-        goto fail;
-    }
-
-    // [hRootKey\OpenRCT2.sv6\shell\open\command]
-    wchar_t szCommandW[MAX_PATH];
-    printResult = swprintf_s(szCommandW, MAX_PATH, L"\"%s\" %s", exePathW, commandArgsW.c_str());
-    assert(printResult >= 0);
-    if (RegSetValueW(hKey, L"shell\\open\\command", REG_SZ, szCommandW, 0) != ERROR_SUCCESS)
-    {
-        goto fail;
-    }
-
-    result = true;
-fail:
-    RegCloseKey(hKey);
-    RegCloseKey(hRootKey);
-    return result;
-}
-
-static void windows_remove_file_association(const utf8* extension)
-{
-#    if _WIN32_WINNT >= 0x0600
-    // [HKEY_CURRENT_USER\Software\Classes]
-    HKEY hRootKey;
-    if (RegOpenKeyW(HKEY_CURRENT_USER, SOFTWARE_CLASSES, &hRootKey) == ERROR_SUCCESS)
-    {
-        // [hRootKey\.ext]
-        RegDeleteTreeA(hRootKey, extension);
-
-        // [hRootKey\OpenRCT2.ext]
-        auto progIdName = get_progIdName(extension);
-        RegDeleteTreeW(hRootKey, progIdName.c_str());
-
-        RegCloseKey(hRootKey);
-    }
-#    endif
-}
-
-void platform_setup_file_associations()
-{
-    // Setup file extensions
-    windows_setup_file_association(".sc4", "RCT1 Scenario (.sc4)", "Play", "\"%1\"", 0);
-    windows_setup_file_association(".sc6", "RCT2 Scenario (.sc6)", "Play", "\"%1\"", 0);
-    windows_setup_file_association(".sv4", "RCT1 Saved Game (.sc4)", "Play", "\"%1\"", 0);
-    windows_setup_file_association(".sv6", "RCT2 Saved Game (.sv6)", "Play", "\"%1\"", 0);
-    windows_setup_file_association(".sv7", "RCT Modified Saved Game (.sv7)", "Play", "\"%1\"", 0);
-    windows_setup_file_association(".td4", "RCT1 Track Design (.td4)", "Install", "\"%1\"", 0);
-    windows_setup_file_association(".td6", "RCT2 Track Design (.td6)", "Install", "\"%1\"", 0);
-
-    // Refresh explorer
-    SHChangeNotify(SHCNE_ASSOCCHANGED, SHCNF_IDLIST, nullptr, nullptr);
-}
-
-void platform_remove_file_associations()
-{
-    // Remove file extensions
-    windows_remove_file_association(".sc4");
-    windows_remove_file_association(".sc6");
-    windows_remove_file_association(".sv4");
-    windows_remove_file_association(".sv6");
-    windows_remove_file_association(".sv7");
-    windows_remove_file_association(".td4");
-    windows_remove_file_association(".td6");
-
-    // Refresh explorer
-    SHChangeNotify(SHCNE_ASSOCCHANGED, SHCNF_IDLIST, nullptr, nullptr);
-}
-
-///////////////////////////////////////////////////////////////////////////////
-
-///////////////////////////////////////////////////////////////////////////////
-// URI protocol association setup
-///////////////////////////////////////////////////////////////////////////////
-
-bool platform_setup_uri_protocol()
-{
-#    if _WIN32_WINNT >= 0x0600
-    log_verbose("Setting up URI protocol...");
-
-    // [HKEY_CURRENT_USER\Software\Classes]
-    HKEY hRootKey;
-    if (RegOpenKeyW(HKEY_CURRENT_USER, SOFTWARE_CLASSES, &hRootKey) == ERROR_SUCCESS)
-    {
-        // [hRootKey\openrct2]
-        HKEY hClassKey;
-        if (RegCreateKeyA(hRootKey, "openrct2", &hClassKey) == ERROR_SUCCESS)
-        {
-            if (RegSetValueA(hClassKey, nullptr, REG_SZ, "URL:openrct2", 0) == ERROR_SUCCESS)
-            {
-                if (RegSetKeyValueA(hClassKey, nullptr, "URL Protocol", REG_SZ, "", 0) == ERROR_SUCCESS)
-                {
-                    // [hRootKey\openrct2\shell\open\command]
-                    wchar_t exePath[MAX_PATH];
-                    GetModuleFileNameW(nullptr, exePath, MAX_PATH);
-
-                    wchar_t buffer[512];
-                    swprintf_s(buffer, std::size(buffer), L"\"%s\" handle-uri \"%%1\"", exePath);
-                    if (RegSetValueW(hClassKey, L"shell\\open\\command", REG_SZ, buffer, 0) == ERROR_SUCCESS)
-                    {
-                        // Not compulsory, but gives the application a nicer name
-                        // [HKEY_CURRENT_USER\SOFTWARE\Classes\Local Settings\Software\Microsoft\Windows\Shell\MuiCache]
-                        HKEY hMuiCacheKey;
-                        if (RegCreateKeyW(hRootKey, MUI_CACHE, &hMuiCacheKey) == ERROR_SUCCESS)
-                        {
-                            swprintf_s(buffer, std::size(buffer), L"%s.FriendlyAppName", exePath);
-                            // mingw-w64 used to define RegSetKeyValueW's signature incorrectly
-                            // You need at least mingw-w64 5.0 including this commit:
-                            //   https://sourceforge.net/p/mingw-w64/mingw-w64/ci/da9341980a4b70be3563ac09b5927539e7da21f7/
-                            RegSetKeyValueW(hMuiCacheKey, nullptr, buffer, REG_SZ, L"OpenRCT2", sizeof(L"OpenRCT2"));
-                        }
-
-                        log_verbose("URI protocol setup successful");
-                        return true;
-                    }
-                }
-            }
-        }
-    }
-#    endif
-
-    log_verbose("URI protocol setup failed");
-    return false;
-}
-
-///////////////////////////////////////////////////////////////////////////////
 #endif
```

I'm not sure removing the large chunk of code was the correct thing to do, but at least I didn't get `The procedure entry point RegDeleteTreeA / RegSetKeyValueA could not be located in the dynamic link library ADVAPI32.dll.` anymore.  

Now I get `This application has failed to start because bcrypt.dll was not found. Re-installing the application may fix this problem.` however.  
Windows Vista and newer have `bcrypt`. WinXP does not and has `wincrypt`.  
I'm creating WinXP compatible FFmpeg binaries from time to time and [restoring the use of `wincrypt`](https://github.com/Reino17/ffmpeg-windows-build-helpers/blob/master/patches/ffmpeg-wincrypt_restore-winxp-compatibility.patch) is THE thing which makes my binaries WinXP compatible again.  
Unlike FFmpeg, I've seen there are many calls to `bcrypt` in `src/openrct2/core/Crypt.CNG.cpp` for instance. In order to create a true WinXP compatible OpenRCT2 binary (I understand now that ReactOS definitely isn't just an open-source WinXP) the use of `wincrypt` needs to be restored, but I doubt anyone of you would want to spend time on this.

Btw, just to be clear, I've been using the `develop`-branch so far.

--- a/src/openrct2-ui/CMakeLists.txt
+++ b/src/openrct2-ui/CMakeLists.txt
@@ -83,6 +83,8 @@ if (WIN32)
         else ()
             find_library(SDL2_LDFLAGS sdl2)
         endif ()
+        # Hardcode some of the libraries used by mingw builds
+        target_link_libraries(${PROJECT} imm32 winmm setupapi version)
     endif ()
 endif ()
 if (MSVC)

--- a/src/openrct2-ui/UiContext.Win32.cpp
+++ b/src/openrct2-ui/UiContext.Win32.cpp
@@ -9,7 +9,7 @@
 
 #ifdef _WIN32
 
-#    ifdef __MINGW32__
+#    if defined(__MINGW32__) && !defined(WINVER) && !defined(_WIN32_WINNT)
 // 0x0600 == vista
 #        define WINVER 0x0600
 #        define _WIN32_WINNT 0x0600

--- a/src/openrct2/CMakeLists.txt
+++ b/src/openrct2/CMakeLists.txt
@@ -13,7 +13,7 @@ if (APPLE)
     set_source_files_properties(${OPENRCT2_CORE_MM_SOURCES} PROPERTIES COMPILE_FLAGS "-x objective-c++ -fmodules")
 endif ()
 
-add_library(${PROJECT_NAME} ${OPENRCT2_CORE_SOURCES} ${OPENRCT2_CORE_MM_SOURCES} ${RCT2_SECTIONS})
+add_library(${PROJECT_NAME} ${OPENRCT2_CORE_SOURCES} ${OPENRCT2_CORE_MM_SOURCES})
 set_target_properties(${PROJECT_NAME} PROPERTIES PREFIX "")
 SET_CHECK_CXX_FLAGS(${PROJECT_NAME})
 
@@ -132,6 +132,11 @@ else ()
                                      ${ZLIB_LIBRARIES}
                                      ${LIBZIP_LIBRARIES})
 endif ()
+
+if (MINGW)
+    # Hardcode libraries used by libzip on mingw
+    target_link_libraries(${PROJECT_NAME} crypto ws2_32)
+endif()
 
 if (UNIX AND NOT ${CMAKE_SYSTEM_NAME} MATCHES "BSD")
     # Include libdl for dlopen

--- a/src/openrct2/core/String.cpp
+++ b/src/openrct2/core/String.cpp
@@ -7,7 +7,7 @@
  * OpenRCT2 is licensed under the GNU General Public License version 3.
  *****************************************************************************/
 
-#ifdef __MINGW32__
+#if defined(__MINGW32__) && !defined(WINVER) && !defined(_WIN32_WINNT)
 // 0x0600 == vista
 #    define WINVER 0x0600
 #    define _WIN32_WINNT 0x0600

--- a/src/openrct2/platform/Platform.Win32.cpp
+++ b/src/openrct2/platform/Platform.Win32.cpp
@@ -22,7 +22,9 @@
 #        define __USE_SHGETKNOWNFOLDERPATH__
 #        define __USE_GETDATEFORMATEX__
 #    else
-#        define ENABLE_VIRTUAL_TERMINAL_PROCESSING 0x0004
+#        ifndef ENABLE_VIRTUAL_TERMINAL_PROCESSING
+#            define ENABLE_VIRTUAL_TERMINAL_PROCESSING 0x0004
+#        endif
 #    endif
 
 #    include "../OpenRCT2.h"

--- a/src/openrct2/platform/Windows.cpp
+++ b/src/openrct2/platform/Windows.cpp
@@ -7,7 +7,7 @@
  * OpenRCT2 is licensed under the GNU General Public License version 3.
  *****************************************************************************/
 
-#ifdef __MINGW32__
+#if defined(__MINGW32__) && !defined(WINVER) && !defined(_WIN32_WINNT)
 // 0x0600 == vista
 #    define WINVER 0x0600
 #    define _WIN32_WINNT 0x0600
@@ -49,6 +49,10 @@
 #    define SINGLE_INSTANCE_MUTEX_NAME "RollerCoaster Tycoon 2_GSKMUTEX"
 
 #    define OPENRCT2_DLL_MODULE_NAME "openrct2.dll"
+
+#    if _WIN32_WINNT < 0x600
+#        define swprintf_s(a, b, c, d, ...) swprintf(a, b, c, ##__VA_ARGS__)
+#    endif
 
 static HMODULE _dllModule = nullptr;
 

